### ISSUE TITLE
Can't refresh treeView<Path>. 

### ISSUE THREAD

**Issue by Kayeeec** [edvin/tornadofx#1307]
Hi, I am implementing a file explorer, which should be able to do basic file operations (create folder, delete folder, moving folders/files by dragging, renaming...). I am having trouble with refreshing the `treeView` on-demand (on button press after a folder is created). 

To repopulate the tree, I have tried to reset the root (with newly created `Path`) and call the `TreeView.refresh()` method, but it is not working. It basically just deletes all the child nodes of the `root` in the `treeView` (so just the root remains without any subfolders shown). 

How can I achieve this? 

I am both TornadoFx/JavaFx beginner and I am in a bit of a hurry. 

Thanks. 

The code (some lines omitted):

```
class FilesExplorerView : View("Working directory") {
    private val controller: FilesExplorerController by inject()
    private val rootFolder = controller.getRootFolder()

    val filesTree = treeview<Path>(TreeItem(rootFolder)) {
        minWidth = 50.0
        populate {
            if (Files.isDirectory(it.value)) it.value.list().filter { path: Path ->
                !Files.isHidden(path) && Files.isReadable(path) && Files.isWritable(path)
            }
            else null
        }
        cellFormat {
            text = "${it.fileName}"
            /* ... omitted ... */
        }
        multiSelect(true)
        root.isExpanded = true
    }

    override val root = vbox {
        hbox {
            alignment = Pos.CENTER_RIGHT
            button("", FontIcon(FontAwesome.REPEAT)) {   /* --- REFRESH BUTTON --- */
                tooltip("Refresh")
                action {

                    // ==== TODO: refresh tree view ???

                    filesTree.root = TreeItem(Paths.get(controller.getRootFolderPathString()))
                    filesTree.refresh()
                }
            }
            button("", FontIcon(FontAwesomeSolid.FOLDER_PLUS)){ /* --- NEW FOLDER BUTTON --- */
                tooltip("New folder")
                action {
                    val scope = Scope()
                    val model = NewFolderModel(NewFolder(rootFolder.toFile().absolutePath, null))
                    setInScope(model, scope)
                    find(NewFolderDialogView::class, scope).openModal(block = true)
                   
                   // ==== TODO: refresh tree view ???

                }
            }
            /* ... other buttons omitted */
        }

        add(filesTree)
        filesTree.fitToParentHeight()
        filesTree.fitToParentWidth()
    }
}

fun Path.list() = Files.list(this).use { it.toList() }
```

BTW using `tornadofx:2.0.0-SNAPSHOT` with Java 11.

Side note - if I have `reloadViewsOnFocus()` in my `MyApp` `init` block (and I run it normally not in debug mode), the tree refreshes itself (of course on focus changes) and shows newly created folders and files correctly. So would it be possible to refresh just the `FilesExplorerView` manually as a workaround? I couldn't find anything about refreshing Views though.

**Comment by Kayeeec** [comment]
Side note - if I have `reloadViewsOnFocus()` in my `MyApp` `init` block, the tree refreshes itself (of course on focus changes) and shows newly created folders and files correctly. So would it be possible to refresh just the `FilesExplorerView` manually as a workaround? I couldn't find anything about refreshing Views though.

**Comment by SchweinchenFuntik** [comment]
NewFolderDialogView code ? 

https://edvin.gitbooks.io/tornadofx-guide/content/part1/5_Data_Controls.html **Data driven TreeView**

**Comment by Kayeeec** [comment]
> NewFolderDialogView code ?
> 
> https://edvin.gitbooks.io/tornadofx-guide/content/part1/5_Data_Controls.html **Data driven TreeView**

`NewFolderDialogView` is just a dialog which takes the name of the new directory and the destination and creates it in the filesystem using `Files.createDirectory(<path compiled from the user inputs>)` (later I will probably replace it by editable TreeCells).

Yes, I looked at the _**Data driven TreeView**_ - I guess I didn't quite understand how to use it in my case - to reflect the directory file structure and react to changes. 

Also, the root folder contents can be changed from outside the app and it also will be synced with a remote repository and multiple files should be movable by selecting and dragging multiple `TreeItems`, which is why I need a robust "refresh" method. 

Well today, as `treeItem.children` already is an  `ObservableList`, I solved it by recursively checking for changes in the directory, adding/removing items as needed:

```
private fun refreshTreeView() {
    with(filesTree){
        refreshTreeItemRecursively(this.root)
    }
}

private fun refreshTreeItemRecursively(root: TreeItem<Path>){
    if (root.value.toFile().isDirectory) {
        val subPathsSet = root.value.list().toSet()
        val toAdd: List<Path> = subPathsSet.filter { path -> !root.children.map { it.value }.toSet().contains(path) }
        val toRemove: List<TreeItem<Path>> = root.children.filter { treeItem -> !subPathsSet.contains(treeItem.value) }
        root.children.removeAll(toRemove)
        root.children.addAll(toAdd.map { TreeItem(it) })
        root.children.forEach { refreshTreeItemRecursively(it) }
    }
}
```

So the `refreshTreeView()` method will be called on _refresh button action_ or whenever needed. This way it preserves expanded items.

Not sure if it could be done better though. 

For user actions where I know where the item was added/removed/modified within the tree, I'll make the single change to the tree manually, everywhere else I'll use the `refreshTreeView` method. 

Perhaps my question was premature, but I felt pretty stuck when I created the issue, as I thought the `TreeView.refresh()` method was supposed to repopulate the treeview ([docs](openjfx.io/javadoc/13/javafx.controls/javafx/scene/control/TreeView.html#re)) and that it could be a bug.

Thanks.
